### ISSUE TITLE
[Feature]: Fog of War Layer

### ISSUE THREAD

**Issue by shondoa** [RPTools/maptool#3447]
### Feature Request

I've been trying to make an effect in my current game with a section of houses with roofs where the interior of the house is only revealed when a player reveals it with their own vision, and otherwise only the roof of the house is shown. I've made a mock-up image here to show what a player should see:

![image](https://user-images.githubusercontent.com/1754184/174418496-ac932991-72a5-43a7-a5b0-d9c076021767.png)

Currently I can only create this effect by putting the interior map on the token layer and the roof map on the background/object layer, however this causes a plethora of other issues, including players not being able to see items on the object layer when the vision is revealed. I'd also prefer if players couldn't click on the map to select it.

### The Solution you'd like

The effect I want could be done with an additional layer type, maybe called the "Fog" layer, which would always visible to players and becomes hidden as they reveal the map. Something similar can already be achieved by assigning a texture to the fog in Edit Map, but the texture loops at the edges and I'd only want a specific portion of the map to have this behavior.

### Alternatives that you've considered.

Originally I thought about requesting an additional layer that has the same vision properties as the Token layer but doesn't allow player interaction, as this would alleviate my current issues, but I think a "Fog" layer would be much easier to implement and have more appealing functionality.

### Additional Context

Currently on 1.10.4.

**Comment by kwvanderlinde** [comment]
This is a really interesting idea!

**Comment by AAUCrisp** [comment]
LOVE this!

**Comment by Phergus** [comment]
Definitely an interesting idea.

FWIW, folks have been accomplishing a similar effect with macros for many years.  I believe the functionality is built-in to the Bag-of-Tricks (BoT) library.  It uses individual roof sections and hides them when tokens enter a building.

**Comment by Jmr3366** [comment]
BoT can't do this.  A token has to move under / close enough to the roof for it (the roof) to fade / disappear.  You're not able to "look in" through a doorway like this.  
Awesome addition if it can be realized.

**Comment by kwvanderlinde** [comment]
I've put together a proof-of-concept for this. It's actually pretty straightforward to implement (assuming the fog layer is part of hard fog, not soft fog).
<img width="820" height="615" alt="Image" src="https://github.com/user-attachments/assets/ee335ac2-9ce8-4610-9cd5-177704884bd7" />

Hard fog defaults to whatever texture is set for it in the map settings. Any stamps or drawings put on the Fog layer will cover it up (or partially cover it, depending on opacity). So in this example, if the roof stamps weren't there on the Fog layer, the hard fog would show as more of that purple texture.

I think the Fog layer would really pop if we modify soft fog so it is just a more transparent version of hard fog. This way, roofs "come back" when a token moves on, while still being a noticably explored part of the map:
<img width="825" height="623" alt="Image" src="https://github.com/user-attachments/assets/3960770d-178f-43bb-9fe7-c1d67abe8532" />

**Comment by FullBleed** [comment]
Interesting.  Is there some way to maybe be able to add some kind of shadow to the revealed portions of the images on the fog layer?  I think it "reads" better.

Note: Mock-up assumes only the roofs are on the fog layer.

<img width="825" height="623" alt="Image" src="https://github.com/user-attachments/assets/48b45a7c-2fe9-41a4-85de-1d57c4dae3cd" />

**Comment by Ryezen** [comment]
This would be an amazing feature!  Glad it to hear it may not take eons of code to implement.  Roofs/cover have always been one of the trickier features to navigate in MT for me.

**Comment by kwvanderlinde** [comment]
> Interesting. Is there some way to maybe be able to add some kind of shadow to the revealed portions of the images on the fog layer? I think it "reads" better.

Yeah, some kind of transition definitely makes that feel better. I'll experiment.

**Comment by kwvanderlinde** [comment]
Alright, here's what I've got for a transition "shadow". It depends on the preference setting for soft fog opacity. 

Here's what it looks like at fog opacity = 120 (same as other screenshots so far):
<img width="322" height="482" alt="Image" src="https://github.com/user-attachments/assets/729e3eb0-bfb6-468e-a05e-7a4eca94f86d" />

For fog opacity = 255, there is no shadow as there is no transition at all:
<img width="323" height="480" alt="Image" src="https://github.com/user-attachments/assets/b66b35b1-e62b-4071-a5c3-8cdf9f152bed" />

For fog opacity = 0, there's a very noticable shadow:
<img width="326" height="486" alt="Image" src="https://github.com/user-attachments/assets/96168e5b-6ccd-4831-808f-0a1b56792aa3" />

**Comment by FullBleed** [comment]
> Alright, here's what I've got for a transition "shadow". It depends on the preference setting for soft fog opacity.

I think the 120 looks pretty good... but I'm going to have to get used to that soft fog and I suspect it won't work well for all maps (like old school blueline maps or just lineart, etc.)

Any chance we could get some soft fog options?  Classic and this new transparency for sure, but maybe some other presets?  I think that if softfog could be desaturated (i.e. black and white) it would actually look really good on most colorful maps.
